### PR TITLE
feat: Add Seed Support for Reproducible Voice Generation (v1 & v2)

### PR DESCRIPTION
**Pull request by elismasilva**
#### Overview

This Pull Request adds random seed support across the VoxCPM generation stack for both v1 and v2 models. The goal is to make speech generation reproducible when a seed is provided, and to expose the actual seed used for a generation so downstream UIs can replay or display it.

---

#### Key Changes

1. **Core models (`VoxCPMModel` and `VoxCPM2Model`)**
   * Added an optional `seed` argument to `_generate` and `_generate_with_prompt_cache`.
   * Added deterministic seed application before each inference attempt.
   * Added `last_successful_seed`, initialized to `None` in both model constructors.
   * When `seed=None`, generation now materializes a concrete seed up front and records it.
   * `last_successful_seed` is updated for every audio-returning path:
     * streaming generation
     * non-streaming generation without badcase retry
     * accepted badcase retry attempts
     * exhausted badcase retries that still return the final attempted audio
   * Badcase retries continue to increment the current seed by `+1` before each retry attempt.

2. **Seed helpers**
   * Added shared helpers in `src/voxcpm/model/utils.py`:
     * `materialize_generation_seed(seed)`
     * `apply_generation_seed(seed)`
   * This keeps v1/v2 seed behavior consistent and avoids duplicated CPU/CUDA RNG setup logic.

3. **High-level pipeline wrapper (`VoxCPM` in `core.py`)**
   * Propagates the `seed` argument through the public `generate` and `generate_streaming` entry points.

4. **Command Line Interface and inference scripts**
   * Added the `--seed` option to the `design`, `clone`, and `batch` subcommands, plus legacy root CLI arguments.
   * Added seed support to the full-finetune and LoRA inference test scripts.

5. **Gradio app**
   * Added a `Seed` field and a `Random Seed` toggle to `app.py`.
   * When `Random Seed` is enabled, the app generates a new seed in the frontend before inference and writes it into the `Seed` field.
   * When `Random Seed` is disabled, users can manually edit the `Seed` field for reproducible generation.
   * After generation completes, the app reads `tts_model.last_successful_seed` and writes the actual successful seed back to the `Seed` field.

6. **Training and validation**
   * Uses a fixed validation seed (`seed=42`) in `generate_sample_audio`, making periodic TensorBoard audio samples easier to compare across training steps.

7. **Tests**
   * Added CLI parser/propagation coverage for `--seed`.
   * Added model utility tests for explicit seed preservation, automatic concrete seed creation when `seed=None`, and CPU/CUDA RNG application.

---

#### How to Test

**Python API:**
```python
wav = model.generate(
    text="This is a test of reproducible voice generation.",
    seed=42,
)
print(f"Last seed: {model.tts_model.last_successful_seed}")
```

**CLI Usage:**
```bash
voxcpm design \
  --text "Reproducible speech synthesis with a fixed seed." \
  --seed 42 \
  --output out.wav
```

When no seed is provided, `last_successful_seed` still records the concrete seed that was generated and used for that run.

**Gradio App:**
* Leave `Random Seed` enabled to generate a fresh seed before each run.
* Disable `Random Seed` to manually reuse or edit the `Seed` value.
* After inference, the `Seed` field is updated with `last_successful_seed`.

### CODE REVIEW

**Comment by a710128** [comment]
Thanks for adding seed support — reproducible generation is a really useful feature, and the propagation through the public API and CLI looks clean. 🙏

I went through the `last_successful_seed` logic in detail and I think it needs a bit more work before it does what the PR description promises. A few related issues around its write paths and initialization:

**1. It's never initialized in `__init__`.**
The attribute is only ever assigned *inside* the generation loop, so if any caller reads `model.tts_model.last_successful_seed` before the first successful generation, it raises `AttributeError` — which is exactly the usage the README example shows. A simple `self.last_successful_seed = None` in `__init__` (both `VoxCPMModel` and `VoxCPM2Model`) would fix this.

**2. Some valid code paths never set it.**
Walking through the branches in `_generate_with_prompt_cache` (and `_generate`):

| Scenario | `last_successful_seed` set? |
|---|---|
| `streaming=True` | ✅ |
| `streaming=False`, `retry_badcase=True`, success | ✅ |
| `streaming=False`, `retry_badcase=False` | ❌ (hits the bare `else: break`) |
| `streaming=False`, `retry_badcase=True`, all retries exhausted | ❌ |

Since `retry_badcase=False` is a perfectly normal option, that path produces audio but leaves `last_successful_seed` stale/unset. It'd be nice to set it on every successful-yield path consistently.

**3. When `seed=None` it stays `None` — which is arguably the most useful case.**
`current_seed = seed`, so if the user doesn't pass a seed (the default), `current_seed` is `None` the whole time and `last_successful_seed` ends up `None`. The scenario where this attribute is most valuable is precisely "I didn't pick a seed, but I want to know which one produced this output so I can reproduce it." To support that, you'd want to materialize a concrete seed up front when `seed is None` (e.g. `current_seed = seed if seed is not None else torch.seed() & 0xFFFFFFFF`), use *that* for `torch.manual_seed`, and record it.

Minor nits while you're in there: a few of these new lines have trailing whitespace, and a couple of the touched scripts dropped their trailing newline (`\ No newline at end of file`).

None of this is hard to address — happy to help review once updated. Thanks again for the contribution! 🚀

**Comment by elismasilva** [comment]
> Thanks for adding seed support — reproducible generation is a really useful feature, and the propagation through the public API and CLI looks clean. 🙏
> 
...

Thanks for the detailed review. I updated the seed handling so `last_successful_seed` now matches the behavior described in the PR.

What changed:

* Initialized `last_successful_seed = None` in both model constructors:
  * `src/voxcpm/model/voxcpm.py`
  * `src/voxcpm/model/voxcpm2.py`
* Added shared seed helpers in `src/voxcpm/model/utils.py`:
  * `materialize_generation_seed(seed)` preserves an explicit seed and creates a concrete seed when `seed=None`.
  * `apply_generation_seed(seed)` applies the seed to PyTorch CPU RNG and CUDA RNG when CUDA is available.
* Updated all four generation paths to use the shared helpers:
  * `VoxCPMModel._generate`
  * `VoxCPMModel._generate_with_prompt_cache`
  * `VoxCPM2Model._generate`
  * `VoxCPM2Model._generate_with_prompt_cache`
* `last_successful_seed` is now written for every path that returns audio:
  * streaming generation
  * non-streaming generation with `retry_badcase=False`
  * successful badcase retry attempts
  * exhausted badcase retries that still return the last attempted audio
* Badcase retries still increment the seed by `+1`, but the recorded value is now the seed from the actual attempt whose audio is returned.
* Added tests for the shared seed helper behavior in `tests/test_model_utils.py`.
* Added the same seed flow to the Gradio app in `app.py`:
  * The UI now has a `Seed` field and a `Random Seed` toggle.
  * With `Random Seed` enabled, the app generates a new seed before inference and writes it into the field.
  * With `Random Seed` disabled, users can edit the seed manually.
  * After generation, the app reads `tts_model.last_successful_seed` and writes the actual successful seed back to the field.

**Review by a710128 — APPROVED** [review]
Thanks for the quick turnaround — this looks great now! 🎉

I re-reviewed the latest commits and all three points are addressed cleanly:

- **Initialization** ✅ `self.last_successful_seed = None` is now set in both `VoxCPMModel` and `VoxCPM2Model.__init__`, so reading it before the first generation is safe.
- **Write-path coverage** ✅ Moving the assignment into the `if not streaming:` block (plus the streaming yield path) means every successful path now records the seed, including `retry_badcase=False` and the exhausted-retries case. The new `last_attempt_seed` tracking also correctly captures the incremented seed after a badcase retry rather than the originally requested one — nice touch.
- **`seed=None` case** ✅ `materialize_generation_seed()` now resolves `None` to a concrete `torch.seed() & 0xFFFFFFFF` value, so `last_successful_seed` finally returns something useful when the user doesn't pass a seed — which was the whole point. Verified it returns an in-range int and preserves explicit seeds.

Also confirmed `tests/test_cli.py` passes (22/22) and the formatting/trailing-whitespace cleanup is in. Extracting the logic into `materialize_generation_seed` / `apply_generation_seed` helpers keeps the call sites tidy too.

Nice work, approving! 🚀

(One optional, non-blocking thought for a future PR: `apply_generation_seed` mutates the process-global torch RNG, so calling `generate(seed=...)` will shift the RNG stream for any unrelated torch code in the caller's process. If you ever want to make it side-effect-free, save/restore the RNG state around generation or thread a local `torch.Generator` through. Totally fine to leave as-is for now.)
